src/mp_integer.hpp

```diff
@@ -830,27 +851,42 @@ struct static_integer
 		}
 		return 0;
 	}
-	// Left-shift by one.
-	void lshift1()
+	// lshift by n bits.
+	int lshift(unsigned long n)
 	{
-		piranha_assert(m_limbs[1u] < (limb_t(1) << (limb_bits - 1u)));
-		using size_type = typename limbs_type::size_type;
-		// Shift both limbs.
-		dlimb_t lo = static_cast<dlimb_t>(static_cast<dlimb_t>(m_limbs[0u]) << dlimb_t(1));
-		dlimb_t hi = static_cast<dlimb_t>(static_cast<dlimb_t>(m_limbs[1u]) << dlimb_t(1));
-		m_limbs[0u] = static_cast<limb_t>(lo);
-		m_limbs[1u] = static_cast<limb_t>(hi + (lo >> limb_bits));
-		piranha_assert((hi >> limb_bits) != 1u);
-		mpz_size_t asize = _mp_size;
+		limb_t hi;
+		limb_t lo;
+		if (n == 0u) return 0;
+		if (m_limbs[0u] == 0u && m_limbs[1u] == 0u) return 0;
+		if (n >= 2u * limb_bits) return 1;
+		if (n >= limb_bits) {
+			if (m_limbs[1u] != 0u) return 1;
+			hi = m_limbs[0u];
+			lo = 0u;
+			if (n == limb_bits) {
+				m_limbs[0u] = hi;
+				m_limbs[1u] = lo;
+				return 0;
+			}
+			n -= limb_bits;
+		} else {
+			hi = m_limbs[1u];
+			lo = m_limbs[0u];
+		}
+
+		if (hi >= (limb_t(1) << (limb_bits - n))) return 1;
+		hi = ulshift(hi, n) + (lo >> limb_bits - n);
```

Comment:
I'd put an explicit cast to ``limb_t`` here: ``hi = static_cast<limb_t>(ulshift(...``. Below it's not necessary as ``ulshift`` already returns the correct type, but here the ``+`` operator could trigger type conversions.